Improve efficiency of at::float32Precision APIs

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164387

I saw this taking up 1% total time when profiling training in torchtitan with (requires Python 3.13+ for PYTHON_PERF_JIT_SUPPORT):

```
MASTER_PORT=3000 MASTER_ADDR=0.0.0.0 RANK=0 LOCAL_RANK=0 WORLD_SIZE=1 PYTHON_PERF_JIT_SUPPORT=1 perf record -m 1024  -D11500 -k 1 -g --call-graph dwarf python -m torchtitan.train --job.config-file ./torchtitan/models/llama3/tra
in_configs/debug_model.toml --compile.enable --training.steps 30 --compile.backend=aot_eager
```

It's a straightforward enough fix (should've used enums all along, IMO) so I went ahead and fixed it.

swolchok committed:
Improve efficiency of at::float32Precision APIs

I saw this taking up 1% total time when profiling training in torchtitan with (requires Python 3.13+ for PYTHON_PERF_JIT_SUPPORT):

```
MASTER_PORT=3000 MASTER_ADDR=0.0.0.0 RANK=0 LOCAL_RANK=0 WORLD_SIZE=1 PYTHON_PERF_JIT_SUPPORT=1 perf record -m 1024  -D11500 -k 1 -g --call-graph dwarf python -m torchtitan.train --job.config-file ./torchtitan/models/llama3/tra
in_configs/debug_model.toml --compile.enable --training.steps 30 --compile.backend=aot_eager
```

It's a straightforward enough fix so I went ahead and fixed it.

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/Context.cpp:
```diff
@@ -40,50 +41,87 @@ namespace {
+  // explicitly rather than loop over an array. If adding more,
+  // consider going back to an array like for operators.
+  detail::Float32Backend result_backend;
+  if (backend == "cuda") {
```
nit: use a switch statement here?

Skylion007 commented on aten/src/ATen/Context.h:
```diff
@@ -488,22 +495,32 @@ class TORCH_API Context {
+  // Yes, it's odd that the inner array has extra length here for the
+  // generic backend, but it seems better than using std::vector.
+  // Map backend->op->precision setting.
+  std::array<
```
Shouldn't these be constexpr or at least const?

benjaminglass1 commented on aten/src/ATen/Context.h:
```diff
@@ -488,22 +495,32 @@ class TORCH_API Context {
+  // Yes, it's odd that the inner array has extra length here for the
+  // generic backend, but it seems better than using std::vector.
+  // Map backend->op->precision setting.
+  std::array<
+      std::pair<
+          detail::Float32Backend,
+          std::
+              array<std::pair<detail::Float32Op, detail::Float32Precision>, 4>>,
+      3>
```
Ugh, valid. Feedback withdrawn!

Skylion007 replied:
@benjaminglass1 It looks like we always index based on the first value of this enum though, could we map the enums to a static offset (like using an int enum) for constant time checking though? That might work, be faster, and more readable, if a bit hacky.

benjaminglass1 replied:
Honestly, based on the amount of `find_if` here...that might be a good idea. I don't love it, but what we have is also a little hacky, so springing for some more readability would be helpful.

swolchok committed:
Update on "Improve efficiency of at::float32Precision APIs"


I saw this taking up 1% total time when profiling training in torchtitan with (requires Python 3.13+ for PYTHON_PERF_JIT_SUPPORT):

```
MASTER_PORT=3000 MASTER_ADDR=0.0.0.0 RANK=0 LOCAL_RANK=0 WORLD_SIZE=1 PYTHON_PERF_JIT_SUPPORT=1 perf record -m 1024  -D11500 -k 1 -g --call-graph dwarf python -m torchtitan.train --job.config-file ./torchtitan/models/llama3/tra
in_configs/debug_model.toml --compile.enable --training.steps 30 --compile.backend=aot_eager
```

It's a straightforward enough fix (should've used enums all along, IMO) so I went ahead and fixed it.

[ghstack-poisoned]

swolchok committed:
make NONE 0 on "Improve efficiency of at::float32Precision APIs"


I saw this taking up 1% total time when profiling training in torchtitan with (requires Python 3.13+ for PYTHON_PERF_JIT_SUPPORT):

```
MASTER_PORT=3000 MASTER_ADDR=0.0.0.0 RANK=0 LOCAL_RANK=0 WORLD_SIZE=1 PYTHON_PERF_JIT_SUPPORT=1 perf record -m 1024  -D11500 -k 1 -g --call-graph dwarf python -m torchtitan.train --job.config-file ./torchtitan/models/llama3/tra
in_configs/debug_model.toml --compile.enable --training.steps 30 --compile.backend=aot_eager
```

It's a straightforward enough fix (should've used enums all along, IMO) so I went ahead and fixed it.

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge
